channel/channel.go

```diff
@@ -322,17 +323,20 @@ func (c *Channel) SignAndAddState(s state.State, sk *[]byte) (state.SignedState,
 
 // UpdateWithChainEvent mutates the receiver if provided with a "new" chain event (with a greater block number than previously seen)
 func (c *Channel) UpdateWithChainEvent(event chainservice.Event) (*Channel, error) {
-	if event.BlockNum() < c.latestBlockNumber {
+	if event.BlockNum() < c.OnChain.LatestBlockNumber {
```

Comment:
We might want to consider _not_ having that logic here, and trusting the chain service to provide "the right kind of events in the right order".

Reply:
>What happens if an event gets replayed (e.g. after a restart)?

Wouldn't it be a case of either the channel has already received the event and it gets ignored (due to `event.Num < LatestBlockNumber`) or the channel hasn't seen the event and updates it's holdings and `LatestBlockNumber`?

Reply:
> We might want to consider not having that logic here, and trusting the chain service to provide "the right kind of events in the right order".

I think events being ordered by block number should be a guarantee of the chain service.  The chain service should always process a block and dispatch all relevant events before moving onto the next block. So I think the events should always be in the "right order" so we shouldn't have to worry about this check having to deal with out of order events.

Reply:
OK I think we can resolve this issue in a follow up PR.

Reply:
https://github.com/statechannels/go-nitro/pull/1591

Reply:
@lalexgap we do have to ensure that events _within the same block_ get processed in the correct order. I can see a footgun where we rely on simply "sorting by block number" and unwittingly permute the order of some events.

Reply:
>@lalexgap we do have to ensure that events within the same block get processed in the correct order. I can see a footgun where we rely on simply "sorting by block number" and unwittingly permute the order of some events.

That's a good point! The chain service should guarantee that it events events based on **transaction order** in the block as well. Since events are stored on transaction receipts in a block this is a pretty natural way of parsing events from a blcok.